[DOCS-15273] Add OpenTelemetry metrics destination (#38656)
DataDog/documentation: content/en/observability_pipelines/destinations/opentelemetry.md

---
title: OpenTelemetry Destination
disable_toc: false
products:
- name: Metrics
  icon: metrics
  url: /observability_pipelines/configuration/?tab=metrics#pipeline-types
---

{{< product-availability >}}

## Overview

> Use Observability Pipelines' OpenTelemetry destination to send metrics over HTTP/S to an OpenTelemetry (OTEL) Collector or another OpenTelemetry Protocol (OTLP)-compatible endpoint.

Reviewer: ```suggestion
Use Observability Pipelines' OpenTelemetry destination to send metrics over HTTP/S to an OpenTelemetry (OTel) Collector or another OpenTelemetry Protocol (OTLP)-compatible endpoint.
```

## Set up destination

<div class="alert alert-danger">For Secrets Management: Only enter the identifier for the HTTP/S Client URI and, if applicable, the TLS key pass. Do <b>not</b> enter the actual values.</div>

Configure the OpenTelemetry destination when you [set up a pipeline][3]. You can set up a pipeline in the [UI][1], using the [API][4], or with [Terraform][5]. The steps in this section are configured in the UI.

After you select the OpenTelemetry destination in the pipeline UI, enter the identifier for your HTTP/S Client URI. If you leave it blank, the [default](#secret-defaults) is used.

**Notes**:

> - The Worker can only send counter, gauge, and histogram metrics to OpenTelemetry. OpenTelemetry does not support other metrics types so the Worker drops them. See [Filter out unsupported metrics](#filter-out-unsupported-metrics) for more information.

Reviewer: ```suggestion
- The Worker can only send counter, gauge, and histogram metrics to OpenTelemetry. OpenTelemetry does not support other metrics types, so the Worker drops them. See [Filter out unsupported metrics](#filter-out-unsupported-metrics) for more information.
```

> - Datadog recommends setting your OTLP receiver to allow out-of-order samples because the Worker doesn't reorder metrics and some OTLP receivers reject out-of-order samples. See [Allow out-of-order samples](#allow-out-of-order-samples) for more information.

Author: Thanks @petere-datadog for the details! I rephrased this part and also the other section, which I renamed to "Allow out-of-order samples".

Let me know what you think.

Reviewer 2: looks great, very concise, thanks.

- If you enter secret identifiers and then choose to use environment variables, the environment variable is the identifier entered and prepended with `DD_OP_`. For example, if you entered `PASSWORD_1` for a password identifier, the environment variable for that password is `DD_OP_PASSWORD_1`.

### Optional settings

#### Enable TLS

{{% observability_pipelines/tls_settings %}}

#### Buffering

{{% observability_pipelines/destination_buffer %}}

## Filter out unsupported metrics

The Worker can only send counter, gauge, and histogram metrics to OpenTelemetry. The following Datadog metrics are not supported because they cannot be converted to OTLP format:

- StatsD-type metrics
- Distribution metrics
- Sketch metrics

If one of these metrics is in a batch to be encoded and sent to OpenTelemetry, the Worker drops the unsupported metric, logs an error, and updates the `component_error_total` metric. Datadog recommends using a [filter processor][9] to filter out unsupported metric types.

> ### Allow out-of-order samples

Author: ```suggestion
## Allow out-of-order samples
```

The Worker doesn't always send metrics in the correct order for a given series because it doesn't reorder metrics. For example, if the first batch of metrics contains metrics with timestamps: `10:03`, `10:04`, `10:05` and the second batch contains metrics with timestamps: `10:01`, `10:02`, `10:06`, the Worker does not reorder those metrics before sending them out.

> Since some OTLP receivers, such as the Prometheus OTLP receiver, rejects out-of-order samples, the second batch of metrics gets rejected by the receiver. As a result, the Worker logs a Bad Request (`400`) error and the entire batch that was rejected gets dropped, even if the OTLP receiver accepted some of the valid metrics in the batch.

Reviewer: ```suggestion
Because some OTLP receivers, such as the Prometheus OTLP receiver, reject out-of-order samples, the second batch of metrics gets rejected by the receiver. As a result, the Worker logs a Bad Request (`400`) error and the entire batch that was rejected gets dropped, even if the OTLP receiver accepted some of the valid metrics in the batch.
```

---
title: OpenTelemetry Destination
disable_toc: false
products:
- name: Metrics
  icon: metrics
  url: /observability_pipelines/configuration/?tab=metrics#pipeline-types
---

{{< product-availability >}}

## Overview

> Use Observability Pipelines' OpenTelemetry destination to send metrics over HTTP/S to an OpenTelemetry (OTel) Collector or other OTLP-compatible endpoint.

Author: ```suggestion
Use Observability Pipelines' OpenTelemetry destination to send metrics over HTTP/S to an OpenTelemetry (OTEL) Collector or other OTLP-compatible endpoint.
```

- If you are sending your metrics to a Prometheus OTLP receiver, set your OTLP receiver to allow out-of order samples. Otherwise, the Prometheus OTLP receiver rejects the out of order sample and the Worker logs a bad request error and drops the batch of metrics.
- If you enter secret identifiers and then choose to use environment variables, the environment variable is the identifier entered and prepended with `DD_OP_`. For example, if you entered `PASSWORD_1` for a password identifier, the environment variable for that password is `DD_OP_PASSWORD_1`.

### Optional settings

#### Enable TLS

{{% observability_pipelines/tls_settings %}}

#### Buffering

{{% observability_pipelines/destination_buffer %}}

## Filter out unsupported metrics

The Worker can only send counter, gauge, and histogram metrics to OpenTelemetry. The following Datadog metrics are not supported because they cannot be converted to OTLP format:

- StatsD-type metrics
- Distribution metrics
- Sketch metrics

> If one of these metrics is in a batch to be encoded and sent to OpenTelemetry, the Worker drops the unsupported metric, logs an error, and updates the `component_error_total` metric. Datadog recommends using a [filter processor][9] to filter out unsupported metric types.

Author: Need to confirm with eng that the unsupported metric gets dropped and not the whole batch.

Reviewer 2: just the unsupported metric.

{{% observability_pipelines/set_secrets_intro %}}

{{< tabs >}}
{{% tab "Secrets Management" %}}

- HTTP/S Client URI endpoint identifier
  - References the HTTP/S URI endpoint to which the Worker sends OpenTelemetry data. For example: `http://localhost:4319/v1/metrics`.
  - The default identifier is `DESTINATION_OTEL_HTTP_CLIENT_URI`.
- HTTP/S Client TLS passphrase identifier (when TLS is enabled):
	- The default identifier is `DESTINATION_OTEL_HTTP_CLIENT_KEY_PASS`.

{{% /tab %}}

{{% tab "Environment Variables" %}}

{{% observability_pipelines/configure_existing_pipelines/destination_env_vars/opentelemetry %}}

{{% /tab %}}
{{< /tabs >}}

## Metrics

For [component metrics][6] and [destination buffer metrics][7] emitted by all destinations, see the [Pipelines Usage Metrics][8] documentation. To filter or group by OpenTelemetry destination metrics, use the tag `component_type:opentelemetry`.

## How the destination works

### Event batching

A batch of events is flushed when one of these conditions occurs. See [event batching][2] for more information.

| Maximum Events | Maximum Size (MB) | Timeout (seconds)   |
|----------------|-------------------|---------------------|

> | TKTK           | TKTK              | TKTK                |

Author: Waiting for eng on this

Reviewer 2: Max events: N/a
Max bytes: 10MB
timeout_secs: 1

  icon: metrics
  url: /observability_pipelines/configuration/?tab=metrics#pipeline-types
---

{{< product-availability >}}

## Overview

Use Observability Pipelines' OpenTelemetry destination to send metrics over HTTP/S to an OpenTelemetry (OTEL) Collector or another OpenTelemetry Protocol (OTLP)-compatible endpoint.

## Set up destination

<div class="alert alert-danger">For Secrets Management: Only enter the identifier for the HTTP/S Client URI and, if applicable, the TLS key pass. Do <b>not</b> enter the actual values.</div>

Configure the OpenTelemetry destination when you [set up a pipeline][3]. You can set up a pipeline in the [UI][1], using the [API][4], or with [Terraform][5]. The steps in this section are configured in the UI.

After you select the OpenTelemetry destination in the pipeline UI, enter the identifier for your HTTP/S Client URI. If you leave it blank, the [default](#secret-defaults) is used.

**Notes**:
- The Worker can only send counter, gauge, and histogram metrics to OpenTelemetry. OpenTelemetry does not support other metrics types so the Worker drops them. See [Filter out unsupported metrics](#filter-out-unsupported-metrics) for more information.

> - If you are sending your metrics to a Prometheus OTLP receiver, set your OTLP receiver to allow out-of-order samples. See [Prometheus OTLP receiver](#prometheus-otlp-receiver) for more information.

Reviewer 2: I think we wanna say that OPW doesn't do any ordering of the metrics, so it can't guarantee that for a given series, it will always send metrics in the correct order therefore if using for example a Prometheus OTLP receiver, we recommend setting it to allow out-of-order samples so OPW doesn't mark the entire metrics batch as dropped b/c of a few out of order metrics. IT should be noted that a prometheus server will try to accept some valid samples in a batch but in OPW we just consume the response and if it's a 400, we assume the entire batch is dropped and update metrics accordingly. 

You can reword this differently, the most important thing is we're calling it out and providing an option to deal with it.

- If you enter secret identifiers and then choose to use environment variables, the environment variable is the identifier entered and prepended with `DD_OP_`. For example, if you entered `PASSWORD_1` for a password identifier, the environment variable for that password is `DD_OP_PASSWORD_1`.

### Optional settings

#### Enable TLS

{{% observability_pipelines/tls_settings %}}

#### Buffering

{{% observability_pipelines/destination_buffer %}}

## Filter out unsupported metrics

The Worker can only send counter, gauge, and histogram metrics to OpenTelemetry. The following Datadog metrics are not supported because they cannot be converted to OTLP format:

- StatsD-type metrics
- Distribution metrics
- Sketch metrics

If one of these metrics is in a batch to be encoded and sent to OpenTelemetry, the Worker drops the unsupported metric, logs an error, and updates the `component_error_total` metric. Datadog recommends using a [filter processor][9] to filter out unsupported metric types.

### Prometheus OTLP receiver

> The Prometheus OTLP receiver rejects out-of-order samples. For example, if the first batch of metrics contains metrics with timestamps: `10:03`, `10:04`, `10:05` and the second batch contains metrics with timestamps: `10:01`, `10:02`, `10:06`, the second batch gets rejected by the Prometheus OTLP receiver. The Worker also logs a Bad Request (`400`) error and the entire second batch gets dropped. To prevent out-of-order samples from getting dropped, you **must** set your OTLP receiver to allow out-of-order samples.

Reviewer 2: let's say the batch gets marked as dropped even though prometheus might have consumed some of the samples in that batch i.e. the in-order samples.
